File: app/src/main/java/appstopper/mobile/cs/fsu/edu/appstopper/WhiteListEntriesAdapter.java
package appstopper.mobile.cs.fsu.edu.appstopper;

import android.arch.lifecycle.ViewModelProviders;
import android.arch.persistence.room.Room;
import android.content.Context;
import android.os.AsyncTask;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.Switch;
import android.widget.TextView;

import java.util.List;

public class WhiteListEntriesAdapter extends RecyclerView.Adapter<WhiteListEntriesAdapter.MyViewHolder> {
    private List<WhitelistEntry> myDataset; //Cached copy of entries
    private final LayoutInflater mInflater;
    private EntryViewModel mEntryViewModel;
    private Context mContext;

    WhiteListEntriesAdapter(Context context, EntryViewModel entryViewModel) {
        mInflater = LayoutInflater.from(context);
        mContext = context;
        mEntryViewModel = entryViewModel;
    }
    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.whitelist_entries_item, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(MyViewHolder holder, final int position) {
        WhitelistEntry current = myDataset.get(position);
        holder.aSwitch.setTag(current.packageName);
        holder.name.setText(current.labelName);
        holder.aSwitch.setOnCheckedChangeListener(null);
        holder.aSwitch.setChecked(current.isWhitelisted);
        holder.aSwitch.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton cView, boolean isChecked) {
                if (isChecked) {
                    mEntryViewModel.setWhitelisted("true", cView.getTag().toString());
                } else {
                    mEntryViewModel.setWhitelisted("false", cView.getTag().toString());
                }
            }
        });
    }
    void setEntries(List<WhitelistEntry> entries) {
        myDataset = entries;
        notifyDataSetChanged();
    }

    // Return the size of your dataset (invoked by the layout manager)
    @Override
    public int getItemCount() {
        if(myDataset != null)
            return myDataset.size();
        else
            return 0;
    }

    class MyViewHolder extends RecyclerView.ViewHolder {
        private final TextView name;
        private final Switch aSwitch;
        private MyViewHolder(View view) {
            super(view);
            name = view.findViewById(R.id.entries_item_name);
            aSwitch = view.findViewById(R.id.entries_item_checkbox);;
        }
    }
}

File: app/src/main/java/WhitelistEntriesAdapter.kt
package appstopper.mobile.cs.fsu.edu.appstopper;

import android.app.PendingIntent.getActivity
import android.arch.persistence.room.Room
import android.content.Context
import android.os.AsyncTask
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.Switch
import android.widget.TextView
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import kotlin.coroutines.coroutineContext

class WhitelistEntriesAdapter (private val myDataSet: ArrayList<WhitelistEntry>, val context: Context):
        RecyclerView.Adapter<WhitelistEntriesAdapter.MyViewHolder>(){
    val db = Room.databaseBuilder(context.applicationContext,
        AppDatabase::class.java, "Whitelist").build()

    class MyViewHolder(val mView: View): RecyclerView.ViewHolder(mView) {
        val name = mView.findViewById(R.id.entries_item_name) as TextView
        val switch = mView.findViewById(R.id.entries_item_checkbox) as Switch
    }

    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): WhitelistEntriesAdapter.MyViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.whitelist_entries_item, parent, false)
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.name.text = myDataSet[position].labelName

        holder.switch.isChecked = myDataSet[position].isWhitelisted
        holder.switch.setOnCheckedChangeListener{ view, isChecked ->
            if (isChecked) {
                myDataSet[position].isWhitelisted = true
                UpdateTask(context).execute(myDataSet[position])
            } else {
                myDataSet[position].isWhitelisted = false
                UpdateTask(context).execute(myDataSet[position])
            }
        }
    }
    override fun getItemCount() = myDataSet.size

    class UpdateTask (val context:Context): AsyncTask<WhitelistEntry, Any, Any>()
    {
        override fun doInBackground(vararg params: WhitelistEntry?) {
            Room.databaseBuilder(context.applicationContext,
                    AppDatabase::class.java, "Whitelist").build().entryDao().updateEntry(params[0])
        }
    }
}